File: 45_JumpGameII.py
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if n == 0 or n == 1:
            return 0
        res = [0] * n
        lastIdx = 1
        for i in range(n):
            if i + nums[i] >= n - 1:
                return res[i] + 1
            for j in range(lastIdx, i + nums[i] + 1):
                res[j] = res[i] + 1
            lastIdx = max(lastIdx, i + nums[i] + 1)
        return -1

File: Greedy/45_JumpGameII.java
public class Solution {
    public int jump(int[] nums) {
        if (nums.length <= 1) {
            return 1;
        }

        int n = nums.length;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, n - 1);
        for (int i = n - 2; n >= 0; n--) {
            int val = n - 1 - i;
            for (int j = 0; j < map.size(); j++) {
                if (map.get(j) - i <= nums[i]) {
                    val = 1 + j;
                }
            }
            map.put(val, i);
            if (i == 0) return val;
        }
        return -1;
    }
}